tests/nn/data_parallel/test_fsdp.py

```diff
@@ -527,6 +527,39 @@ def _test_transformer(self, rank, group, config):
         assert objects_are_equal(ref_output, no_grad_output, raise_exception=True)
 
 
+class TestBasicTrainAndEvalWithCheckpointing(DistributedTest):
+    def test_train_and_eval_with_checkpointing(self):
+        #  test_fn = functools.partial(self._test_train_and_eval_with_checkpointing)
+        test_fn = self._test_train_and_eval_with_checkpointing
+        spawn_and_init(test_fn)
+
+    @classmethod
+    def _test_train_and_eval_with_checkpointing(self, rank, group):
+        # Keep initialization deterministic.
+        torch.manual_seed(0)
+        model = FullyShardedDataParallel(SimpleModuleWithCheckpointing().cuda())
+        optim = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
+
+        # Train for a step.
+        torch.manual_seed(1 + rank)
+        input = torch.randn(2, 3).cuda()
+        loss = model(input).sum()
+        loss.backward()
+        optim.step()
+
+        # Now do an eval step.
+        model.eval()
+        with torch.no_grad():
+            input = torch.randn(2, 3).cuda()
+            model(input).sum()
```

Comment:
after this line, you need to put the model back in train mode, right?

Reply:
Ah yes, good catch. Still get the same error though.

I just did some debugging and everything looks fine until the final `loss.backward()` is called. Right before this, all sharded (flattened) parameters are in the correct shape. The inner sharded and checkpointed module has shape `(6,)`. But after the final `loss.backward()` call, this parameter now has shape `(12,)`, and so the gradient has shape `(12,)` which causes the error in the `optim.step()` call, since the optimizer is expecting a gradient of size `(6,)` for the momentum update.

Reply:
hmm, that's very intriguing! I think it might be related to the `checkpoint_fwd_counter` value maintained by checkpoint_wrapper. Can you check that value is correct (0) after the eval case?

Reply:
Ah, yea this value is 1 after the eval case.

Reply:
Gotcha, I'll try that.

Reply:
Hmm, it seems that `torch.is_grad_enabled()` never returns `True` within `CheckointFunction.forward()`. As a consequence, `check_backward_validity()` is never ran.

Reply:
Is there any reason not to stick with my original solution of just calling `original_forward` directly when gradients aren't required?